xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/page/events/PageNotificationEventListener.java

```diff
@@ -0,0 +1,155 @@
+        {
+            // If the event is expected by our descriptor
+            if (descriptor.getEventTriggers().contains(event.getClass().getCanonicalName())
+                    && checkXObjectCondition(descriptor, source)
+                    && this.evaluateVelocityTemplate(descriptor.getAuthorReference(),
+                    descriptor.getValidationExpression())) {
+                observationManager.notify(
+                        new PageNotificationEvent(descriptor),
+                        "org.xwiki.platform:xwiki-platform-notifications-api", source);
+            }
+        }
+    }
+
+    /**
+     * Ensure that the given the source matches what the descriptor needs.
+     * If the source is an instance of XWikiDocument, will check if the document contains the XObject specified in
+     * the descriptor.
+     *
+     * @param descriptor the event descriptor
+     * @param source the event source
+     * @return
+     */
+    private boolean checkXObjectCondition(PageNotificationEventDescriptor descriptor, Object source) {
+        if (StringUtils.isBlank(descriptor.getObjectType())) {
+            // If no XObject is specified in the Object Type field, we don’t need to check the source.
+            return true;
+        } else if (source instanceof XWikiDocument) {
+            XWikiDocument document = (XWikiDocument) source;
+            Map<DocumentReference, List<BaseObject>> documentXObjects = document.getXObjects();
+            /*  We can’t create a DocumentReference when only using descriptor.objectType, so we will have to
+                    iterate through the map */
+            for (DocumentReference documentReference : documentXObjects.keySet())  {
+                if (documentReference.toString().equals(descriptor.getObjectType())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Evaluate the given velocity template and return a boolean.
+     *
+     * @param userReference a user reference used to build context
+     * @param templateContent the velocity template that should be evaluated
+     * @return true if the template evaluation returned «true» or if the template is empty
+     */
+    private boolean evaluateVelocityTemplate(DocumentReference userReference, String templateContent)
+    {
+        try {
+            // We don’t need to evaluate the template if it’s empty
+            if (StringUtils.isBlank(templateContent)) {
+                return true;
+            }
+
+            Template customTemplate = new StringTemplate(
+                    templateContent,
+                    Syntax.XWIKI_2_1,
+                    Syntax.PLAIN_1_0,
+                    userReference);
+
+            StringWriter writer = new StringWriter();
+            templateManager.render(customTemplate, writer);
+
+            return writer.toString().trim().contains("true");
```

Comment:
It’s `contains()` because the template evaluation can return elements such as `<p>`, `</p>` or `&nbsp`.